### PR TITLE
[docs] Add WASM interface C++ reference documentation

### PR DESCRIPTION
**Pull request by victorj8**
## Change Description

Currently there is few documentation about host functions on EOS, this PR provides documentation for some host functions.

## Change Type
**Select ONE**
- [x] Documentation

- [ ] Stability bug fix

- [ ] Other

- [ ] Other - special case

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [x] Documentation Additions

### CODE REVIEW

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp
Comment (lparisc) [inline libraries/chain/include/eosio/chain/webassembly/interface.hpp:265]: This table is not rendering correctly. Check https://www.doxygen.nl/manual/tables.html

File: docs.json
Comment (lparisc) [inline docs.json:70]: Need to add the "doxybook" generator in "docs.json" (after the "doxygen_to_xml" generator) for the contents to render on the left navigation menu:
```
            ...
        },
        {
            "name": "doxybook",
            "options": {}
        }
```

**Review by lparisc — CHANGES_REQUESTED** [review]

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp
Comment (spoonincode) [inline libraries/chain/include/eosio/chain/webassembly/interface.hpp:378]: It may be worth noting somehow that this assertion does not require the EOSIO canonical check to pass. That is, a signature that would have been disallowed during transaction acceptance would potentially be allowed here. Many contract users probably don't care about that detail.. but it's an important subtle behavior of the interface

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp
Comment (spoonincode) [inline libraries/chain/include/eosio/chain/webassembly/interface.hpp:612]: Is this correct?

**Comment by spoonincode** [comment]
Should the documentation offer any guidance or recommendations on usage patterns of any of these? For example, should the documentation recommend to use the assert_sha256() variant over sha256() if contract's usage pattern allows it. Or, avoiding spamage of eosio_assert() with negative checks.

File: docs.json
Comment (victorj8) [reply docs.json:70]: done

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp
Comment (victorj8) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp:612]: there is an exception being thrown 

`EOS_THROW( eosio_assert_message_exception, "assertion failure with message: ${s}", ("s",message) );`

but correct me if this is not the case or it requires a different wording

Comment (spoonincode) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp:612]: I think the description is negated? Isn't the action aborted when the condition is _false_

Comment (victorj8) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp]: that's right, thanks for catching that.

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp
Comment (victorj8) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp:378]: I am somewhat out of my element here, but if you have a suggested text i can add it, let me know what text should i add.

File: libraries/chain/include/eosio/chain/webassembly/interface.hpp
Comment (victorj8) [reply libraries/chain/include/eosio/chain/webassembly/interface.hpp:266]: Thanks for the observation, but after trying the table solution the html table still was not rendering properly

/**
 * Gets the maximum key size, maximum value size, and maximum iterators of a kv database and returns the size of the data.
 * The kv parameters are encoded as 16 bytes, representing four 32-bit little-endian values.
 * <table>
 * <tr><th>byte <th> 0 | 1 | 2 | 3 <th> 4 | 5 | 6 | 7 <th> 8 | 9 | 10 | 11 <th> 12 | 13 | 14 | 15
 * <tr><td>field <td>version <td>key limit <td>value limit <td>max iterators
 * <tr><td>type <td>0 <td>32-bits LE <td>32-bits LE <td>32-bits LE
 * </table>
 *
 * @ingroup privileged
 * @param[out] packed_kv_parameters - the buffer containing the packed kv parameters.
 * @param max_version - has no effect, but should be 0.
 *
 * @return Returns the size required in the buffer (if the buffer is too small, nothing is written).
 *
*/

I switched to use codeblocks instead.

**Comment by lparisc** [comment]
> Should the documentation offer any guidance or recommendations on usage patterns of any of these? For example, should the documentation recommend to use the assert_sha256() variant over sha256() if contract's usage pattern allows it. Or, avoiding spamage of eosio_assert() with negative checks.

Yes, of course. If the usage suggestions are short and to the point they can be added in a "remarks" section within the applicable function. However, I believe a separate "overview" section that includes best practices is also desired for devs to be able to use the WASM interface correctly and effectively. To that end, I have created this issue #9518 to keep this effort active. In the meantime I'm approving this unless there are any hard corrections. @victorj8 please wait at least 24 hours before merging in case there are any additional corrections. Thanks

**Review by lparisc — APPROVED** [review]

**Pull request merged**
